index.rst

```diff
@@ -1,29 +1,481 @@
+
+The main disadvantage of this scheme is that using existence checks to poll for metadata and log datasets for completed results in more load on the storage system.
+WMS-specific tooling wrapped by BPS might be able to do this more efficiently.
+If not, load could still be mitigated by increasing the fraction of the time the job spends sleeping, and possibly ensuring that these jobs go to a special queue that throttles their overall activity (including overall database insert load).
+
+Assuming the special batch-scheduling behavior needed for this job is not difficult to implement in BPS, it probably provides the best balance between load-control and ease of implementation.
+
+User-space Daemon
+-----------------
+
+It is not uncommon for BPS users to run ``bps report`` repeatedly while their workflows are running.
+This suggests that users might appreciate being able to just run the aggregation tool (in sleep/poll) mode themselves, since the tool would thus naturally be able to provide them with the status of both the execution and the ingestion of outputs into the butler (with former slightly out-of-date).
+It is also much easier to diagnose problems in aggregation when they occur directly in a user's interactive shell rather than a batch worker.
+
+This scheme would be incompatible with having a ``finalJob`` that also runs aggregation, unless we introduce a locking mechanism, so it would rely on users remembering to actually run the tool and keeping it running over the course of a workflow.
+As with a dedicated DAG job, load from existence-check polling may also be a concern if the sleep fraction is not high enough, and in this case there would be no coordinated way to throttle multiple workflows from multiple users.
+
+Given these disadvantages, this would probably not be our preferred long-term approach for production jobs, but it would be extremely useful to have available for development and debugging.
+
+Multi-workflow Aggregation Service
+----------------------------------
+
+To provide the most control over how database inserts are batched together and fully minimize existence checks, we could stand up a data-facility-wide service that receives messages (e.g. Kafka) from completed batch jobs to trigger aggregation.
+Such a service would need to be notified when a new submission is started, allowing it to set up a new aggregation graph for it, and it would append to each graph only when either a certain number of quanta completed or a certain amount of time had passed.
+It would need to provide ways for users to cancel active workflows and determine when a workflow's outputs had been fully aggregated and ingested into the butler.
+
+The main downside of this approach is that it is significantly more complex in code development and especially deployment than other options.
```

Comment:
On the other hand the observatory has *extensive* experience with Kafka through the EFD, prompt processing, and alert generation so it's not hard to bring in expertise to set up the broker.

Reply:
I think there might still be `finalJob` in this scheme, but it would still only send Kafka messages to its server, and I'm not sure we could make its completion meaningful unless it had a way to receive a "workflow complete" message back from the service.